src/System.Management.Automation/engine/PSVersionInfo.cs

```diff
@@ -571,77 +647,142 @@ public static bool TryParse(string version, out SemanticVersion result)
                 return false;
             }
 
-            result._parsedVersion = dashIndex < 0
-                ? new SemanticVersion(major, minor, patch)
-                : new SemanticVersion(major, minor, patch, version.Substring(dashIndex + 1));
-            return true;
-        }
-
-        private static bool TryParseComponent(string component, string componentName, ref VersionResult result, out int parsedComponent)
-        {
-            if (!Int32.TryParse(component, NumberStyles.Integer, CultureInfo.InvariantCulture, out parsedComponent))
+            if (match.Groups["patch"].Success && !int.TryParse(match.Groups["patch"].Value, out patch))
             {
-                result.SetFailure(ParseFailureKind.FormatException, component);
+                result.SetFailure(ParseFailureKind.FormatException);
                 return false;
             }
 
-            if (parsedComponent < 0)
+            if (preLabel != null && !Regex.IsMatch(preLabel, s_LabelUnitRegEx) ||
+               (buildLabel != null && !Regex.IsMatch(buildLabel, s_LabelUnitRegEx)))
             {
-                result.SetFailure(ParseFailureKind.ArgumentOutOfRangeException, componentName);
+                result.SetFailure(ParseFailureKind.FormatException);
                 return false;
             }
 
+            result._parsedVersion = new SemanticVersion(major, minor, patch, preLabel, buildLabel);
             return true;
         }
 
         /// <summary>
-        /// ToString
+        /// Implement ToString()
         /// </summary>
         public override string ToString()
         {
-            if (Patch < 0)
+            if (versionString == null)
             {
-                return string.IsNullOrEmpty(Label)
-                    ? StringUtil.Format("{0}.{1}", Major, Minor)
-                    : StringUtil.Format("{0}.{1}-{2}", Major, Minor, Label);
+                StringBuilder result = new StringBuilder();
+
+                result.Append(Major).Append(Utils.Separators.Dot).Append(Minor).Append(Utils.Separators.Dot).Append(Patch);
+
+                if (!string.IsNullOrEmpty(PreLabel))
+                {
+                    result.Append("-").Append(PreLabel);
+                }
+
+                if (!string.IsNullOrEmpty(BuildLabel))
+                {
+                    result.Append("+").Append(BuildLabel);
+                }
+
+                versionString = result.ToString();
             }
 
-            return string.IsNullOrEmpty(Label)
-                ? StringUtil.Format("{0}.{1}.{2}", Major, Minor, Patch)
-                : StringUtil.Format("{0}.{1}.{2}-{3}", Major, Minor, Patch, Label);
+            return versionString;
+        }
+
+        /// <summary>
+        /// Implement Compare.
+        /// </summary>
+        public static int Compare(SemanticVersion versionA, SemanticVersion versionB)
+        {
+            if (versionA == null) return -1;
```

Comment:
what about versionA and versionB are both null?

Reply:
Good catch!

Fixed.